examples/span-inspector/src/inspector.rs

```diff
@@ -0,0 +1,299 @@
+
+        #[qinvokable]
+        #[cxx_name = "rebuildOutput"]
+        fn rebuild_output(self: Pin<&mut SpanInspector>, pos: i32);
+    }
+
+    impl UniquePtr<QTextDocument> {}
+    impl cxx_qt::Threading for SpanInspector {}
+}
+
+use cxx_qt::{CxxQtType, Threading};
+use qobject::{QQuickTextDocument, QString, QTextDocument};
+use std::{pin::Pin, ptr};
+
+pub struct SpanInspectorRust {
+    cursor_position: Arc<Mutex<i32>>,
+    input: *mut QQuickTextDocument,
+    output: *mut QQuickTextDocument,
+}
+
+impl Default for SpanInspectorRust {
+    fn default() -> Self {
+        Self {
+            cursor_position: Arc::new(Mutex::new(3)),
+            input: ptr::null_mut(),
+            output: ptr::null_mut(),
+        }
+    }
+}
+
+impl qobject::SpanInspector {
+    unsafe fn output_document(&self) -> Pin<&mut QTextDocument> {
+        if self.output.is_null() {
+            panic!("Output document must be set!");
+        }
+        let output = unsafe { &*self.output };
+        unsafe { Pin::new_unchecked(&mut *output.text_document()) }
+    }
+
+    fn set_input(mut self: Pin<&mut Self>, input: *mut QQuickTextDocument) {
+        self.as_mut().rust_mut().input = input;
+        self.as_mut().input_changed();
+    }
+
+    fn rebuild_output(self: Pin<&mut Self>, pos: i32) {
+        let mut cursor_position = match self.cursor_position.lock() {
+            Ok(mutex) => mutex,
+            Err(poison_error) => poison_error.into_inner(),
+        };
+        *cursor_position = pos;
+
+        let input = unsafe { Pin::new_unchecked(&mut *self.input) };
+        let cursor_position = self.cursor_position.clone();
+        let qt_thread = self.qt_thread();
+        qt_thread
+            .queue(|this| {
+                unsafe { this.output_document() }
+                    .set_html(&QString::from(String::from("expanding...")))
+            })
+            .ok();
+
+        let text = unsafe { Pin::new_unchecked(&mut *input.text_document()) }
+            .to_plain_text()
+            .to_string();
+
+        std::thread::spawn(move || {
+            let cursor_position = match cursor_position.lock() {
+                Ok(mutex) => mutex,
+                Err(poison_error) => poison_error.into_inner(),
+            };
+
+            let output = match Self::expand(&text, *cursor_position as usize) {
+                Ok((expanded, span_data)) => {
+                    let Ok(file) = syn::parse_file(expanded.as_str())
+                        .map_err(|err| eprintln!("Parsing error: {err}"))
+                    else {
+                        return;
+                    };
+                    Self::build_html(&prettyplease::unparse(&file), span_data)
+                }
+                Err(error) => Self::build_error_html(&error),
+            };
+            qt_thread
+                .queue(move |this| {
+                    unsafe { this.output_document() }.set_html(&QString::from(output))
```

Comment:
If you do multiple subsequent edits to the input string, they all race to set their result as the output_string. If a later edit finishes faster, it will be overwritten by an edit that happened beforehand, leaving the app in an inconsistent state.

The easiest way to fix this is to add a `request_token` (just an i32) that is incremented every time a background thread is started.
Only the background thread with the matching request_token is then allowed to set its result.

Reply:
@QuentinLeeWeber Once this is fixed, this is ready for approval.